assets/mdk/brittleness.mdk

```diff
@@ -0,0 +1,179 @@
+Colorizer     : dafnyx
+
+Heading Depth: 0
+
+Css Header:
+  body.madoko {
+    max-width: 100%;
+    margin: 0;
+    padding: 0;
+  }
+  .page-content {
+    padding: 0;
+  }
+
+# What is brittleness?
+
+Dafny is designed to integrate programming and verification. You can write both programs and specification in Dafny, using annotations such as preconditions and postconditions on implementation code. To ensure that the code meets the specification, you may also need to annotate the code with the outline of a proof, such as including an invariant on a loop. Some amount of annotation can improve your code, providing precise documentation on what makes the code work. However, too much annotation would is distracting, especially if it amounts to trivialities. To make this work, Dafny relies on automated theorem proving using an SMT solver. These solvers are effective and allow for much of the verification effort to be performed automatically. However, SMT-based automation comes at a cost. If you rely too much on automation, you may find that the outcome of the automated verification becomes chaotic. A postcondition of a method for which the SMT solver used to be able to find a proof may fail to verify for seemingly unrelated changes such as upgrading to a new version of Dafny, adding an unrelated definition to your development, or even changing the name of a variable. This issue is sometimes referred to as *brittleness*. While we make every effort to push the boundary of how much can be automatically verified with Dafny, a serious Dafny user must learn to write annotations to prevent brittleness. This can be done thanks to the application of common software engineering principles such as modularity and abstraction, and also by structuring proof annotations to guide the SMT solver more precisely.
+
+ In this post, we will dig into how to avoid brittleness in practice, what properties of a program tend to push it near this problematic threshold, and how to write Dafny code that consistently verifies. To begin, let’s consider an example program that exhibits the problem. We’ll start by defining a type of rational numbers. 
```

Comment:
Perhaps we should clarify that the example is designed specifically to exacerbate the problem, and that one doesn't run into brittleness so easily in general?